src/Primitives/arduino.cpp

```diff
@@ -1026,3 +1074,44 @@ bool resolve_external_memory(char *symbol, Memory **val) {
     FATAL("Could not find memory %s \n", symbol);
     return false;
 }
+
+//------------------------------------------------------
+// Restore external state when restoring a snapshot
+//------------------------------------------------------
+void restore_external_state(Module *m,
+                            std::vector<IOStateElement> external_state) {
+    uint8_t opcode = *m->pc_ptr;
+    // TODO: Maybe primitives can also be called using the other call
+    // instructions such as call_indirect
+    //  maybe there should just be a function that checks if a certain function
+    //  is being called that handles all these cases?
+    if (opcode == 0x10) {  // call opcode
+        uint8_t *pc_copy = m->pc_ptr + 1;
+        uint32_t fidx = read_LEB_32(&pc_copy);
+        if (fidx < m->import_count) {
+            for (auto &primitive : primitives) {
+                if (!strcmp(primitive.name, m->functions[fidx].import_field)) {
+                    if (primitive.f_reverse) {
+                        debug("Reversing action for primitive %s\n",
+                              primitive.name);
+                        primitive.f_reverse(m, external_state);
+                    }
+                    return;
+                }
+            }
+        }
+    }
+}
+
+//------------------------------------------------------
+// Serialize external state into a snapshot
+//------------------------------------------------------
+std::vector<IOStateElement *> get_io_state(Module *m) {
+    std::vector<IOStateElement *> ioState;
+    for (auto &primitive : primitives) {
+        if (primitive.f_serialize_state) {
+            primitive.f_serialize_state(ioState);
+        }
+    }
+    return ioState;
+}
```

Comment:
Again, this seems to be the same code across platforms, so should be moved to another file to avoid code duplication.

Reply:
Ah I hadn't thought of this. Okay, for now that's a good enough reason for me not to refactor this.

But we should keep this issue in the foreground of our minds, so it doesn't stay on the main branch too long ... 😅

Reply:
I added an issue #257.